language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala

```diff
@@ -68,12 +69,40 @@ class DottyLanguageServer extends LanguageServer
         val classpathFlags = List("-classpath", (config.classDirectory +: config.dependencyClasspath).mkString(File.pathSeparator))
         val sourcepathFlags = List("-sourcepath", config.sourceDirectories.mkString(File.pathSeparator), "-scansource")
         val settings = defaultFlags ++ config.compilerArguments.toList ++ classpathFlags ++ sourcepathFlags
-        myDrivers.put(config, new InteractiveDriver(settings))
+        myDrivers(config) = new InteractiveDriver(settings)
       }
     }
     myDrivers
   }
 
+  /** Restart all presentation compiler drivers, copying open files over */
+  private def restart() = thisServer.synchronized {
+    interactiv.println("restarting presentation compiler")
+    val driverConfigs = for ((config, driver) <- myDrivers.toList) yield
+      (config, new InteractiveDriver(driver.settings), driver.openedFiles)
+    for ((config, driver, _) <- driverConfigs)
+      myDrivers(config) = driver
+    System.gc()
+    for ((_, driver, opened) <- driverConfigs; (uri, source) <- opened)
+      driver.run(uri, source)
+    if (Memory.isCritical())
+      println(s"WARNING: Insufficient memory to run Scala language server on these projects.")
+  }
+
+  private def checkMemory() =
+    if (Memory.isCritical())
+      CompletableFutures.computeAsync { _ => restart(); new Object() }
+        // new Object() necessary or we get a BootstrapMethodError:
+        //
+        // Caused by: java.lang.invoke.LambdaConversionException: Type mismatch for lambda expected return: void is not convertible to class java.lang.Object
+        // at java.lang.invoke.AbstractValidatingLambdaMetafactory.validateMetafactoryArgs(AbstractValidatingLambdaMetafactory.java:286)
+        // at java.lang.invoke.LambdaMetafactory.metafactory(LambdaMetafactory.java:303)
+        // at java.lang.invoke.CallSite.makeSite(CallSite.java:302)
+        // ... 11 more
+        //
+        // This looks like a problem with Dottys code generation for void-returning closures passed
+        // to Java methods. (or SAM functions in general?)
```

Comment:
Hmm we should be returning a BoxedUnit here, not sure could you open an issue so we can keep track of this?

Reply:
See #3984

Reply:
Yes, you are right.